src/dag.rs

```diff
@@ -202,69 +146,101 @@ impl<G: GraphNode> Dag<G> {
-        // https://codingots.medium.com/tree-traversal-without-recursion-221cbea6d004
-        // This is a translation of that, where "s1" is temp_stack and s2 is
-        // self.visit_stack. In that version there is only one root node but
-        // here we have several.
-        // First phase is to build up the stack of nodes to visit using
-        // temp_stack as an intermediate.
-        if self.visit_stack.is_empty() {
-            let mut temp_stack = vec![self.unvisited_roots.pop()?];
-            while let Some(cur_idx) = temp_stack.pop() {
-                self.visit_stack.push(cur_idx);
-                for child_idx in &self.dag.edges[cur_idx] {
-                    temp_stack.push(*child_idx);
-                }
-            }
+impl<'a, G: GraphNode> TopologicalSort<'a, G> {
+    fn new(dag: &'a Dag<G>, sources: Vec<usize>) -> Self {
+        TopologicalSort {
+            dag,
+            dfs_stack: sources,
+            node_state: vec![NodeState::New; dag.nodes.len()],
+            node_in_cycle: None,
         }
-        // Now we just cruise down the to_visit stack drinking a large bottle of
-        // cranberry juice listening to Fleetwood Mac.
-        Some(&self.dag.nodes[self.visit_stack.pop().unwrap()])
     }
-}
 
-#[derive(Clone)]
-pub struct TopDown<'a, G: GraphNode> {
-    dag: &'a Dag<G>,
-    visit_stack: Vec<usize>,
-    // Note these "roots" don't have to be roots in the overall DAG, they are
-    // only roots of the top-down traversals we're doing.
-    unvisited_roots: Vec<usize>,
+    fn check_cycles(mut self) -> Option<usize> {
+        while self.next().is_some() {}
+        self.node_in_cycle
+    }
 }
 
-impl<'a, G: GraphNode> Iterator for TopDown<'a, G> {
+impl<'a, G: GraphNode> Iterator for TopologicalSort<'a, G> {
     type Item = &'a G;
 
-    fn next(&mut self) -> Option<&'a G> {
-        if self.visit_stack.is_empty() {
-            self.visit_stack.push(self.unvisited_roots.pop()?)
-        }
-        let cur_idx = self.visit_stack.pop().unwrap();
-        for child_idx in &self.dag.edges[cur_idx] {
-            self.visit_stack.push(*child_idx);
+    // This is the iterative version of the DFS-based toposort implementation
+    // (https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search).
+    // Basically: do a normal DFS starting from the sources, and a node is
+    // appended to the toposort as soon as it is closed.
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(v) = self.dfs_stack.pop() {
+            match self.node_state[v] {
+                NodeState::New => {
+                    self.node_state[v] = NodeState::Opened;
+                    self.dfs_stack.push(v); // Don't actually pop yet.
+                                            // If any child of v is `Opened`, we fonud a cycle!
+                                            // This includes the case of a self-loop at v.
```

Comment:
Looks like rustfmt got a bit, er, creative here - did you mean for "Don't actually pop yet" to refer to the `push` call and the next part to refer to the cycle check?

Reply:
Lol yeah, had to give the first comment its own line or it wouldn't be happy. :/